smolyay/basis.py

```diff
@@ -0,0 +1,226 @@
+            self._compute_points()
+        return self._points
+
+    @property
+    def n(self):
+        """degree of polynomial"""
+        return self._n
+
+    @n.setter
+    def n(self,n):
+        if self._n != n:
+            self._n = n
+            self._points = []
+
+    def _compute_points(self):
+        """Compute extrema of Chebyshev polynomial of the first kind"""
+        if self._n == 0:
+            self._points = [0]
+        else:
+            i = numpy.arange(self._n+1)
+            self._points = list(-numpy.cos(numpy.pi*i/self._n))
+
+    def __call__(self,x):
+        """Terms of basis function
+        Returns the output of an nth degree Chebyshev polynomial of the
+        first kind with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        if self._n == 0:
+            return 1
+        elif self._n == 1:
+            return x
+        else:
+            k_lim = self._n//2
+            answer = 0
+            for k in range(0,k_lim+1):
+                answer += math.comb(self._n,2*k)*((x**2 - 1)**k)*(x**(self._n-2*k))
+            return answer
+
+
+class BasisFunctionSet():
+    """Set of basis functions
+    Creates a framework for a set of basis functions to be used as the
+    building blocks for a surrogate model.
+    ``basis_set`` is a list of BasisFunction objects
+    ''all_points'' 1D points taken from the BasisFunction objects
+
+    """
+
+    def __init__(self,all_points,basis_set):
+        self._basis_set = basis_set
+        self._all_points = all_points
+
+    @property
+    def all_points(self):
+        """All points for basis function set at some level of precision"""
+        return self._all_points
+
+    @all_points.setter
+    def all_points(self,all_points):
+        self._all_points = all_points
+
+    @property
+    def basis_set(self):
+        """list of BasisFunction objects"""
+        return self._basis_set
+
+    @basis_set.setter
+    def basis_set(self,basis_set):
+        self._basis_set = basis_set
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
+    """Set of basis functions that are nested
+
+    The set of basis functions this class uses are assumed to be part
+    of a sequential series of functions, where the functions that
+    determine the points are nested
+    ``levels`` is a list of lists that determines which grid level each
+    point is added on
+    """
+
+    def __init__(self,all_points,basis_set,levels):
+        """Initialization of parameters"""
+        super().__init__(all_points,basis_set)
+        self._levels = levels
+
+    @property
+    def levels(self):
+        """list of list that show the points for each grid level"""
+        return self._levels
+
+    @levels.setter
+    def levels(self,levels):
+        self._levels = levels
+
+
+def make_nested_chebyshev_points(exactness,basis_function):
```

Comment:
This is nice! I would make it a classmethod of the Chebyshev basis function called `make_nested_set` that takes only one argument `exactness`, currently.

Reply:
I'm not sure what you mean by "classmethod of the Chebyshev basis function" but I can do the rest.

Reply:
I mean this function should be a classmethod for ChebyshevFirstKind:

```
@classmethod
def make_nested_set(cls, exactness):
    …
```